app/models/package_manager/go/go_mod_parser.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+class PackageManager::Go
+  class GoModParser
+    attr_reader :mod_contents
+
+    RETRACT_WITH_PARENS = /^\s*retract\s\((.*?)\)/m.freeze
+    RETRACT_WITHOUT_PARENS = /retract\s(.+)/.freeze
+    MODULE_REGEX = /module\s+(.+)/.freeze
+
+    def initialize(mod_contents)
+      @mod_contents = mod_contents
+    end
+
+    # Scans contents for retract directives parses their contents
+    #
+    # @return [Array<String, Array<String, String>>] an array consisting of:
+    #   a version number (String) or
+    #   a range represented by a two element array: [range_start_number, range_end_number]
+    def retracted_version_ranges
+      return @retracted_version_ranges if @retracted_version_ranges
+
+      retract_specs = []
+      retract_specs.concat stripped_contents.scan(RETRACT_WITH_PARENS).flatten
+      retract_specs.concat stripped_contents.scan(RETRACT_WITHOUT_PARENS).flatten
+
+      @retracted_version_ranges = retract_specs.flat_map do |retract_spec|
+        parse_retract(retract_spec)
+      end
+    end
+
+    def canonical_module_name
+      module_line = stripped_contents.lines.find { |line| line.match(MODULE_REGEX) }
```

Comment:
If you `return` early here if `module_line` is `nil`, you can skip needing all of the existence operators.